Split state storage interface into 3 parts

wallyworld opened the pull request:
## Description of change

The state storage interface is split into 3 separate, smaller interfaces:
- storage instance management
- volume storage
- file storage

The affected facades are storagecommon, uniter, modelmanager, application. Also, storagecommon contained watcher code that was only used by the uniter facade so this was moved across.

Followup PRs will refactor IAAS and CAAS model types to add supported storage features to both models.

## QA steps

Deploy postgresql charm with storage
attach/detach/import storage etc

babbageclunk commented on apiserver/common/storagecommon/storage.go:
```diff
@@ -314,22 +275,30 @@ func UnitStorage(st StorageInterface, unit names.UnitTag) ([]state.StorageInstan
 // removed. Any storage that is not found will be omitted.
 func ClassifyDetachedStorage(
-	st StorageInterface,
+	st StorageInstanceInterface,
```
Same here.

babbageclunk commented on apiserver/facades/agent/uniter/state.go:
```diff
@@ -10,45 +10,79 @@ import (
 
-var getStorageState = func(st *state.State) (storageStateInterface, error) {
-	im, err := st.IAASModel()
+var getStorageState = func(st *state.State) (storageInterface, storageVolumeInterface, storageFilesystemInterface, error) {
```
I think my suggestion would tidy this up too - you'd return one thing + error from here rather than four.

wallyworld replied:
Yep, fixed this

babbageclunk commented on apiserver/facades/agent/uniter/storage.go:
```diff
@@ -406,3 +415,54 @@ func accessUnitTag(tag string, canAccess func(names.Tag) bool) (names.UnitTag, e
+// to the VolumeAttachmentInfo or FilesystemAttachmentInfo corresponding to the
+// tags specified.
+func watchStorageAttachment(
+	st storageInterface,
```
Same here.

babbageclunk commented on apiserver/facades/agent/uniter/storage.go:
```diff
@@ -406,3 +415,54 @@ func accessUnitTag(tag string, canAccess func(names.Tag) bool) (names.UnitTag, e
+	var watchers []state.NotifyWatcher
+	switch storageInstance.Kind() {
+	case state.StorageKindBlock:
+		volume, err := stVolume.StorageInstanceVolume(storageTag)
```
Don't you need to check for nil stVolume here?

wallyworld replied:
done

babbageclunk commented on apiserver/facades/agent/uniter/storage.go:
```diff
@@ -406,3 +415,54 @@ func accessUnitTag(tag string, canAccess func(names.Tag) bool) (names.UnitTag, e
+			stVolume.WatchBlockDevices(machineTag),
+		}
+	case state.StorageKindFilesystem:
+		filesystem, err := stFile.StorageInstanceFilesystem(storageTag)
```
Couldn't stFile be nil here? (I think maybe this couldn't happen, but it's probably not safe to assume it here.)

wallyworld replied:
done

babbageclunk commented on apiserver/facades/client/application/application.go:
```diff
@@ -60,10 +60,16 @@ type APIv6 struct {
 type APIBase struct {
-	backend    Backend
+	backend  Backend
+	stVolume storagecommon.StorageVolumeInterface
```
Poor old StorageInstanceInterface, not as popular as the other interfaces.

babbageclunk commented on apiserver/facades/client/application/application_unit_test.go:
```diff
@@ -47,9 +47,13 @@ var _ = gc.Suite(&ApplicationSuite{})
 func (s *ApplicationSuite) setAPIUser(c *gc.C, user names.UserTag) {
 	s.authorizer.Tag = user
 	api, err := application.NewAPIBase(
+		&s.backend,
```
\<shudders\>

babbageclunk commented on apiserver/facades/client/storage/shim.go:
```diff
@@ -66,11 +69,14 @@ func NewFacadeV3(
+		registry, pm, resources, authorizer)
 }
 
 type storageAccess interface {
```
Actually, this naming isn't too bad either (if you don't like Instances, Filesystems, Volumes): `InstanceAccess, FilesystemAccess, VolumeAccess`?

anastasiamac replied:
+1 I like the 'access' suffix here...

wallyworld commented on apiserver/common/storagecommon/storage.go:
```diff
@@ -72,7 +66,9 @@ type StorageInterface interface {
 // by a charm.
 func StorageAttachmentInfo(
-	st StorageInterface,
+	st StorageInstanceInterface,
```
I need to have the 3 interfaces passed in here because callers embed the core interfaces in sub interfaces and if these are placed on a common super interfaces with getters, the sigs don't match

wallyworld commented on apiserver/common/storagecommon/storage.go:
```diff
@@ -9,61 +9,55 @@ import (
-// StorageInterface is an interface for obtaining information about storage
-// instances and related entities.
-type StorageInterface interface {
+// StorageInstanceInterface is an interface for obtaining information about storage
```
Yeah, I think you are right. I went with the suggestion to use InstanceAccess

wallyworld commented:
I've fixed the interface names and reworked the facades, especially the application and model manager ones, to fully separate storage from state. The facades no longer have constructors which take 3 storage args. But 2 of the common storage methods still need the 3 args for interface compatibility.

wallyworld commented:
$$merge$$
